src/cargo/util_schemas/manifest.rs

```diff
@@ -1108,27 +1110,130 @@ impl TomlTarget {
     }
 }
 
+macro_rules! str_newtype {
+    ($name:ident) => {
+        #[derive(Serialize, Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+        #[serde(transparent)]
+        pub struct $name<T: AsRef<str> = String>(T);
+
+        impl<T: AsRef<str>> $name<T> {
+            pub fn into_inner(self) -> T {
+                self.0
+            }
+        }
+
+        impl<T: AsRef<str>> std::convert::AsRef<str> for $name<T> {
+            fn as_ref(&self) -> &str {
+                self.0.as_ref()
+            }
+        }
+
+        impl<T: AsRef<str>> std::ops::Deref for $name<T> {
+            type Target = T;
+
+            fn deref(&self) -> &Self::Target {
+                &self.0
+            }
+        }
+
+        impl<T: AsRef<str>> std::borrow::Borrow<str> for $name<T> {
+            fn borrow(&self) -> &str {
+                self.0.as_ref()
+            }
+        }
+
+        impl<'a> std::str::FromStr for $name<String> {
+            type Err = anyhow::Error;
+
+            fn from_str(value: &str) -> Result<Self, Self::Err> {
+                Self::new(value.to_owned())
+            }
+        }
+
+        impl<'de, T: AsRef<str> + serde::Deserialize<'de>> serde::Deserialize<'de> for $name<T> {
+            fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+            where
+                D: serde::Deserializer<'de>,
+            {
+                let inner = T::deserialize(deserializer)?;
+                Self::new(inner).map_err(serde::de::Error::custom)
+            }
+        }
+
+        impl<T: AsRef<str>> Display for $name<T> {
+            fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+                self.0.as_ref().fmt(f)
+            }
+        }
+    };
+}
+
+str_newtype!(PackageName);
+
+impl<T: AsRef<str>> PackageName<T> {
+    pub fn new(name: T) -> Result<Self> {
```

Comment:
Could we have some doc like this?

```suggestion  
    /// Validates if input is a valid package name, and returns a wrapper type of it.
    pub fn new(name: T) -> Result<Self> {
```

Or for all `str_newtype` we could have comment like

```rust
/// Wrapper type that checks the input package name in valid during the creation.
```